### PR TITLE
[WIP] Error logging callback

### PR DESCRIPTION
**Pull request by bmosaicml**
# What does this PR do?
This PR adds a callback that logs ICL outputs during eval. It modifies the custom metrics to keep track of incorrect model outputs. Each metric is responsible for specifying the table schema for logging cached responses as well as specifying how to format the cached responses using the tokenizer.

The EvalOutputLogging callback is then responsible for logging the cached results in table format after each evaluation.

Design [doc](https://www.notion.so/Eval-output-logging-callback-cdf475b61c4a4cf6b2b65d18910380af)

# What issue(s) does this change relate to?

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/metrics/nlp.py [hunk 1]
```diff
@@ -270,19 +326,31 @@ def update(self, outputs: List[str], labels: List[List[str]], batch: Optional[Di
...
 
             cleaned_final_answer = self.normalize_answer(final_answer)
             cleaned_sample_labels = {self.normalize_answer(label) for label in sample_labels}
-
+            correct = False
             if any(cleaned_final_answer.startswith(label) for label in cleaned_sample_labels):
                 self.correct += torch.tensor(1.0)
+                correct = True
+
+            assert isinstance(self.response_cache, list)
```
Comment (maxisawesome) [inline composer/metrics/nlp.py:341]: Add `if self.cache_responses` here?

**Comment by maxisawesome** [comment]
Looks good. Would be nice if there was an option to save results to a file or the cloud but that might be better suited for a different PR doing it for all evals

**Comment by bmosaicml** [comment]
> Looks good except for my one comment. Would be nice if there was an option to save results to a file or the cloud but that might be better suited for a different PR doing it for all evals

I believe this will work out of the box as composer's logger functionality will include wandb/s3/etc in the logger destinations and then log them all simultaneously. I am going to mock up a test branch on foundry that tests s3 and wandb!

File: composer/metrics/nlp.py (hunk 1 above)
Comment (bmosaicml) [reply composer/metrics/nlp.py:342]: response_cache will be a list regardless since it is part of the super class's ctor

File: composer/callbacks/eval_output_logging_callback.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    def _write_tables_to_output_dir(self, state: State):
+        try:
+            import pandas as pd
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='pandas',
+                                                conda_package='pandas',
+                                                conda_channel='conda-forge') from e
+        # write tmp files
+        self.hash.update((str(time.time()) + str(random.randint(0, 1_000_000))).encode('utf-8'))
+        tmp_dir = os.getcwd() + '/' + self.hash.hexdigest()
```
Comment (maxisawesome) [inline composer/callbacks/eval_output_logging_callback.py:70]: This seems overcomplicated lol just use tempfile?

File: composer/callbacks/eval_output_logging_callback.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        self.table = {}
+
+    def eval_end(self, state: State, logger: Logger) -> None:
+        # eval start runs after each benchmark's evaluator
+        # during each eval, only a single dataloader/benchmark will be active
+        assert state.dataloader is not None
+        assert isinstance(state.dataloader, DataLoader)
+        if hasattr(state.dataloader, 'dataset') and isinstance(state.dataloader.dataset, ICLDatasetTypes):
+            assert isinstance(state.dataloader.dataset, ICLDatasetTypes)
+            if hasattr(state.dataloader.dataset, 'tokenizer'):
```
Comment (maxisawesome) [inline composer/callbacks/eval_output_logging_callback.py:128]: Do datasets ever not have tokenizers?

File: composer/metrics/nlp.py [hunk 4]
```diff
@@ -196,6 +196,37 @@ def compute(self) -> Tensor:
...
+        if self.cache_responses and len(self.response_cache) > 0:
+            rows = []
+            for row in self.response_cache:
+                assert isinstance(row, dict)
+                columns = list(row.keys())
+                converted_row = []
+                for r_i in row.values():
+                    if isinstance(r_i, list) and len(r_i) > 0 \
+                        and all(isinstance(r_ij, int) for r_ij in r_i) \
+                        and not all(isinstance(r_ij, bool) for r_ij in r_i):
```
Comment (maxisawesome) [inline composer/metrics/nlp.py:220]: I don't understand why we need all this type checking.

File: composer/callbacks/eval_output_logging_callback.py (hunk 3 above)
Comment (bmosaicml) [reply composer/callbacks/eval_output_logging_callback.py:128]: They always do, but this is necessary for pyright to pass

File: composer/metrics/nlp.py (hunk 4 above)
Comment (bmosaicml) [reply composer/metrics/nlp.py:220]: see prev comment re pyright

**Comment by maxisawesome** [comment]
Successfully merged [here](https://github.com/mosaicml/composer/pull/2977)
